File: kotlin/src/test/kotlin/com/edgeveda/sdk/ModelRegistryTest.kt
package com.edgeveda.sdk

import org.junit.Assert.*
import org.junit.Test

/**
 * Unit tests for ModelRegistry.
 *
 * Verifies model counts, query functions, ModelType assignments, and ID uniqueness.
 * Mirrors the Flutter gold standard ModelRegistry group in edge_veda_test.dart.
 */
class ModelRegistryTest {

    // ── Model counts ──────────────────────────────────────────────────────────

    @Test
    fun `getAllTextModels returns 5 models`() =
        assertEquals(5, ModelRegistry.getAllTextModels().size)

    @Test
    fun `getVisionModels returns 1 model`() =
        assertEquals(1, ModelRegistry.getVisionModels().size)

    @Test
    fun `getWhisperModels returns 2 models`() =
        assertEquals(2, ModelRegistry.getWhisperModels().size)

    @Test
    fun `getEmbeddingModels returns 1 model`() =
        assertEquals(1, ModelRegistry.getEmbeddingModels().size)

    @Test
    fun `total model count across all categories is 10`() {
        val all = ModelRegistry.getAllTextModels() +
                  ModelRegistry.getVisionModels() +
                  listOf(ModelRegistry.smolvlm2_500m_mmproj) +
                  ModelRegistry.getWhisperModels() +
                  ModelRegistry.getEmbeddingModels()
        assertEquals(10, all.size)
    }

    // ── getModelById ──────────────────────────────────────────────────────────

    @Test
    fun `getModelById returns model for known text model id`() {
        val model = ModelRegistry.getModelById("llama-3.2-1b-instruct-q4")
        assertNotNull(model)
        assertEquals("Llama 3.2 1B Instruct", model!!.name)
    }

    @Test
    fun `getModelById returns model for known whisper model id`() {
        val model = ModelRegistry.getModelById("whisper-tiny-en")
        assertNotNull(model)
        assertEquals(ModelType.WHISPER, model!!.modelType)
    }

    @Test
    fun `getModelById returns model for vision model id`() {
        val model = ModelRegistry.getModelById("smolvlm2-500m-video-instruct-q8")
        assertNotNull(model)
        assertEquals(ModelType.VISION, model!!.modelType)
    }

    @Test
    fun `getModelById returns model for mmproj id`() {
        val model = ModelRegistry.getModelById("smolvlm2-500m-mmproj-f16")
        assertNotNull(model)
        assertEquals(ModelType.MMPROJ, model!!.modelType)
    }

    @Test
    fun `getModelById returns model for embedding model id`() {
        val model = ModelRegistry.getModelById("all-minilm-l6-v2-f16")
        assertNotNull(model)
        assertEquals(ModelType.EMBEDDING, model!!.modelType)
    }

    @Test
    fun `getModelById returns null for unknown id`() =
        assertNull(ModelRegistry.getModelById("does-not-exist"))

    @Test
    fun `getModelById returns null for empty string`() =
        assertNull(ModelRegistry.getModelById(""))

    // ── getMmprojForModel ─────────────────────────────────────────────────────

    @Test
    fun `getMmprojForModel returns projector for smolvlm2_500m`() {
        val mmproj = ModelRegistry.getMmprojForModel(ModelRegistry.smolvlm2_500m.id)
        assertNotNull(mmproj)
        assertEquals(ModelType.MMPROJ, mmproj!!.modelType)
    }

    @Test
    fun `getMmprojForModel returns null for text model`() {
        val mmproj = ModelRegistry.getMmprojForModel(ModelRegistry.llama32_1b.id)
        assertNull(mmproj)
    }

    @Test
    fun `getMmprojForModel returns null for unknown id`() =
        assertNull(ModelRegistry.getMmprojForModel("unknown-model-id"))

    // ── ModelType assignments ─────────────────────────────────────────────────

    @Test
    fun `llama32_1b has TEXT model type`() =
        assertEquals(ModelType.TEXT, ModelRegistry.llama32_1b.modelType)

    @Test
    fun `phi35_mini has TEXT model type`() =
        assertEquals(ModelType.TEXT, ModelRegistry.phi35_mini.modelType)

    @Test
    fun `gemma2_2b has TEXT model type`() =
        assertEquals(ModelType.TEXT, ModelRegistry.gemma2_2b.modelType)

    @Test
    fun `tinyLlama has TEXT model type`() =
        assertEquals(ModelType.TEXT, ModelRegistry.tinyLlama.modelType)

    @Test
    fun `qwen3_06b has TEXT model type`() =
        assertEquals(ModelType.TEXT, ModelRegistry.qwen3_06b.modelType)

    @Test
    fun `smolvlm2_500m has VISION model type`() =
        assertEquals(ModelType.VISION, ModelRegistry.smolvlm2_500m.modelType)

    @Test
    fun `smolvlm2_500m_mmproj has MMPROJ model type`() =
        assertEquals(ModelType.MMPROJ, ModelRegistry.smolvlm2_500m_mmproj.modelType)

    @Test
    fun `whisperTinyEn has WHISPER model type`() =
        assertEquals(ModelType.WHISPER, ModelRegistry.whisperTinyEn.modelType)

    @Test
    fun `whisperBaseEn has WHISPER model type`() =
        assertEquals(ModelType.WHISPER, ModelRegistry.whisperBaseEn.modelType)

    @Test
    fun `allMiniLmL6V2 has EMBEDDING model type`() =
        assertEquals(ModelType.EMBEDDING, ModelRegistry.allMiniLmL6V2.modelType)

    // ── Model descriptor invariants ───────────────────────────────────────────

    @Test
    fun `all models have non-empty ids`() {
        val all = ModelRegistry.getAllTextModels() +
                  ModelRegistry.getVisionModels() +
                  listOf(ModelRegistry.smolvlm2_500m_mmproj) +
                  ModelRegistry.getWhisperModels() +
                  ModelRegistry.getEmbeddingModels()
        all.forEach { model ->
            assertTrue("Model '${model.name}' has empty id", model.id.isNotEmpty())
        }
    }

    @Test
    fun `all models have non-empty names`() {
        val all = ModelRegistry.getAllTextModels() +
                  ModelRegistry.getVisionModels() +
                  listOf(ModelRegistry.smolvlm2_500m_mmproj) +
                  ModelRegistry.getWhisperModels() +
                  ModelRegistry.getEmbeddingModels()
        all.forEach { model ->
            assertTrue("Model '${model.id}' has empty name", model.name.isNotEmpty())
        }
    }

    @Test
    fun `all models have positive sizeBytes`() {
        val all = ModelRegistry.getAllTextModels() +
                  ModelRegistry.getVisionModels() +
                  listOf(ModelRegistry.smolvlm2_500m_mmproj) +
                  ModelRegistry.getWhisperModels() +
                  ModelRegistry.getEmbeddingModels()
        all.forEach { model ->
            assertTrue("Model '${model.id}' sizeBytes must be > 0", model.sizeBytes > 0)
        }
    }

    @Test
    fun `all models have non-empty downloadUrl`() {
        val all = ModelRegistry.getAllTextModels() +
                  ModelRegistry.getVisionModels() +
                  listOf(ModelRegistry.smolvlm2_500m_mmproj) +
                  ModelRegistry.getWhisperModels() +
                  ModelRegistry.getEmbeddingModels()
        all.forEach { model ->
            assertTrue("Model '${model.id}' has empty downloadUrl", model.downloadUrl.isNotEmpty())
        }
    }

    @Test
    fun `all model ids are unique`() {
        val all = ModelRegistry.getAllTextModels() +
                  ModelRegistry.getVisionModels() +
                  listOf(ModelRegistry.smolvlm2_500m_mmproj) +
                  ModelRegistry.getWhisperModels() +
                  ModelRegistry.getEmbeddingModels()
        val uniqueIds = all.map { it.id }.toSet()
        assertEquals(
            "Duplicate model IDs found",
            all.size,
            uniqueIds.size
        )
    }

    @Test
    fun `all text models are not included in whisper or embedding query results`() {
        val textIds = ModelRegistry.getAllTextModels().map { it.id }.toSet()
        val whisperIds = ModelRegistry.getWhisperModels().map { it.id }.toSet()
        val embeddingIds = ModelRegistry.getEmbeddingModels().map { it.id }.toSet()
        assertTrue((textIds intersect whisperIds).isEmpty())
        assertTrue((textIds intersect embeddingIds).isEmpty())
    }

    // ── Registry model identity ────────────────────────────────────────────────

    @Test
    fun `llama32_1b id matches expected string`() =
        assertEquals("llama-3.2-1b-instruct-q4", ModelRegistry.llama32_1b.id)

    @Test
    fun `whisperTinyEn id matches expected string`() =
        assertEquals("whisper-tiny-en", ModelRegistry.whisperTinyEn.id)

    @Test
    fun `whisperBaseEn id matches expected string`() =
        assertEquals("whisper-base-en", ModelRegistry.whisperBaseEn.id)

    @Test
    fun `smolvlm2_500m id matches expected string`() =
        assertEquals("smolvlm2-500m-video-instruct-q8", ModelRegistry.smolvlm2_500m.id)

    @Test
    fun `allMiniLmL6V2 id matches expected string`() =
        assertEquals("all-minilm-l6-v2-f16", ModelRegistry.allMiniLmL6V2.id)
}

File: web/src/__tests__/ModelRegistry.test.ts
import { ModelRegistry } from '../ModelRegistry';

describe('ModelRegistry', () => {
  describe('model counts', () => {
    it('getAllTextModels returns 5 text models', () => {
      expect(ModelRegistry.getAllTextModels()).toHaveLength(5);
    });

    it('getVisionModels returns 1 vision model (not mmproj)', () => {
      expect(ModelRegistry.getVisionModels()).toHaveLength(1);
    });

    it('getWhisperModels returns 2 whisper models', () => {
      expect(ModelRegistry.getWhisperModels()).toHaveLength(2);
    });

    it('getEmbeddingModels returns 1 embedding model', () => {
      expect(ModelRegistry.getEmbeddingModels()).toHaveLength(1);
    });

    it('getImageModels returns 1 image generation model', () => {
      expect(ModelRegistry.getImageModels()).toHaveLength(1);
    });

    it('total across all categories is 11 (includes mmproj and image gen)', () => {
      // 5 text + 1 vision + 1 mmproj + 2 whisper + 1 embedding + 1 image = 11
      expect(ModelRegistry.getAllModels()).toHaveLength(11);
    });
  });

  describe('getModelById', () => {
    it('finds llama-3.2-1b-instruct-q4 by ID', () => {
      const model = ModelRegistry.getModelById('llama-3.2-1b-instruct-q4');
      expect(model).not.toBeNull();
      expect(model!.id).toBe('llama-3.2-1b-instruct-q4');
    });

    it('llama model name contains "Llama"', () => {
      const model = ModelRegistry.getModelById('llama-3.2-1b-instruct-q4');
      expect(model!.name).toContain('Llama');
    });

    it('returns null for unknown ID', () => {
      expect(ModelRegistry.getModelById('does-not-exist')).toBeNull();
    });

    it('finds smolvlm2 mmproj by ID', () => {
      const model = ModelRegistry.getModelById('smolvlm2-500m-mmproj-f16');
      expect(model).not.toBeNull();
      expect(model!.modelType).toBe('mmproj');
    });
  });

  describe('getMmprojForModel', () => {
    it('returns mmproj for smolvlm2-500m-video-instruct-q8', () => {
      const mmproj = ModelRegistry.getMmprojForModel('smolvlm2-500m-video-instruct-q8');
      expect(mmproj).not.toBeNull();
      expect(mmproj!.modelType).toBe('mmproj');
    });

    it('returns null for a text model', () => {
      expect(ModelRegistry.getMmprojForModel('llama-3.2-1b-instruct-q4')).toBeNull();
    });

    it('returns null for an unknown model', () => {
      expect(ModelRegistry.getMmprojForModel('unknown-model')).toBeNull();
    });
  });

  describe('data integrity', () => {
    it('all model IDs are unique across all categories', () => {
      const allModels = ModelRegistry.getAllModels();
      const ids = allModels.map((m) => m.id);
      const uniqueIds = new Set(ids);
      expect(uniqueIds.size).toBe(ids.length);
    });

    it('all models have positive sizeBytes', () => {
      ModelRegistry.getAllModels().forEach((model) => {
        expect(model.sizeBytes).toBeGreaterThan(0);
      });
    });

    it('all downloadUrls start with https://', () => {
      ModelRegistry.getAllModels().forEach((model) => {
        expect(model.downloadUrl).toMatch(/^https:\/\//);
      });
    });

    it('all models have non-empty id and name', () => {
      ModelRegistry.getAllModels().forEach((model) => {
        expect(model.id.trim()).not.toBe('');
        expect(model.name.trim()).not.toBe('');
      });
    });

    it('all models have a non-empty format', () => {
      ModelRegistry.getAllModels().forEach((model) => {
        expect(model.format.trim()).not.toBe('');
      });
    });
  });

  describe('model types and properties', () => {
    it('whisperTinyEn.modelType === "whisper"', () => {
      expect(ModelRegistry.whisperTinyEn.modelType).toBe('whisper');
    });

    it('whisperBaseEn.modelType === "whisper"', () => {
      expect(ModelRegistry.whisperBaseEn.modelType).toBe('whisper');
    });

    it('allMiniLmL6V2.modelType === "embedding"', () => {
      expect(ModelRegistry.allMiniLmL6V2.modelType).toBe('embedding');
    });

    it('smolvlm2_500m_mmproj.format === "GGUF"', () => {
      expect(ModelRegistry.smolvlm2_500m_mmproj.format).toBe('GGUF');
    });

    it('smolvlm2_500m_mmproj.modelType === "mmproj"', () => {
      expect(ModelRegistry.smolvlm2_500m_mmproj.modelType).toBe('mmproj');
    });

    it('llama32_1b.downloadUrl starts with https://', () => {
      expect(ModelRegistry.llama32_1b.downloadUrl).toMatch(/^https:\/\//);
    });

    it('qwen3_06b.modelType === "text"', () => {
      expect(ModelRegistry.qwen3_06b.modelType).toBe('text');
    });

    it('vision model (smolvlm2_500m) is not in getAllTextModels', () => {
      const textIds = ModelRegistry.getAllTextModels().map((m) => m.id);
      expect(textIds).not.toContain('smolvlm2-500m-video-instruct-q8');
    });

    it('mmproj is not in getVisionModels', () => {
      const visionIds = ModelRegistry.getVisionModels().map((m) => m.id);
      expect(visionIds).not.toContain('smolvlm2-500m-mmproj-f16');
    });

    it('sdV21Turbo.modelType === "imageGeneration"', () => {
      expect(ModelRegistry.sdV21Turbo.modelType).toBe('imageGeneration');
    });

    it('sdV21Turbo is findable by ID', () => {
      const model = ModelRegistry.getModelById('sd-v2-1-turbo-q8');
      expect(model).not.toBeNull();
      expect(model!.name).toContain('SD');
    });
  });
});
